src/bin/server.rs

```diff
@@ -246,73 +264,113 @@ fn load_config(config_file: &str) -> (SocketAddr, Vec<u8>) {
-            break;
+    let mut merkle = MerkleTree::new();
+    let mut requests = Vec::with_capacity(batch_size as usize);
+
+    macro_rules! check_ctrlc {
+        () => {
+            if !keep_running.load(Ordering::Acquire) {
+                warn!("Ctrl-C caught, exiting...");
+                return;
+            }
         }
+    }
+
+    loop {
+        check_ctrlc!();
 
         poll.poll(&mut events, poll_duration).expect("poll failed");
 
         for event in events.iter() {
+
             match event.token() {
                 MESSAGE => {
-                    loop {
-                        match socket.recv_from(&mut buf) {
-                            Ok((num_bytes, src_addr)) => {
-                                if let Ok(nonce) = nonce_from_request(&buf, num_bytes) {
-                                    let resp = make_response(&mut ephemeral_key, cert_bytes, nonce);
-                                    let resp_bytes = resp.encode().unwrap();
-
-                                    let bytes_sent = socket.send_to(&resp_bytes, &src_addr).expect("send_to failed");
-
-                                    num_responses += 1;
-                                    info!("Responded {} bytes to {} for '{}..' (resp #{})", bytes_sent, src_addr, hex::encode(&nonce[0..4]), num_responses);
-                                } else {
-                                    num_bad_requests += 1;
-                                    info!("Invalid request ({} bytes) from {} (resp #{})", num_bytes, src_addr, num_responses);
+
+                    let mut done = false;
+
+                    'process_batch: loop {
+                        check_ctrlc!();
+
+                        merkle.reset();
+                        requests.clear();
+
+                        let resp_start = NUM_RESPONSES.load(Ordering::SeqCst);
+
+                        for i in 0..batch_size {
+                            match socket.recv_from(&mut buf) {
+                                Ok((num_bytes, src_addr)) => {
+                                    if let Ok(nonce) = nonce_from_request(&buf, num_bytes) {
+                                        requests.push((Vec::from(nonce), src_addr));
+                                        merkle.push_leaf(nonce);
+                                    } else {
+                                        num_bad_requests += 1;
+                                        info!("Invalid request ({} bytes) from {} (#{} in batch, resp #{})", num_bytes, src_addr, i, resp_start + i as usize);
+                                    }
+                                },
+                                Err(e) => match e.kind() {
+                                    ErrorKind::WouldBlock => {
+                                        done = true;
+                                        break;
+                                    },
+                                    _ => panic!("recv_from failed with {:?}", e)
```

Comment:
Being a long-lived server, the `panic!` here makes me go hmmm. 

Could `recv_from` return a "benign" error like `EINTR`? Why not log it and continue?

I'm guessing that I don't fully understand the contract `mio` is providing here w.r.t. potential errors.

Reply:
That's a much better idea - I'll change it to do what master currently does.